Add standard for DNS

markus-hentsch opened the pull request.

horazont commented on Standards/scs-01xx-v1-dns.md:
```diff
@@ -0,0 +1,122 @@
+A local DNS recursor can be used to cache and serve DNS responses locally. It servers as a proxy between the clients and external DNS servers.
+This improves performance and speed of DNS resolution in the infrastructure.
+Furthermore, it can be configured to use DNSSEC, DNS over HTTPS and/or DNS over TLS to increase security and privacy of DNS requests it handles for clients.
+Even if individual clients do not support these functionalities, they can still benefit from a local recursor's implementation of those and don't have to trust external DNS servers directly.
```
I think it is worth spelling out explicitly that this is the only way to get "zero configuration" DNSSEC validation and benefits for all clients. Even clients which have no clue that DNSSEC or DoT even exists are protected properly if a local validating recursor is provided by the CSP.

markus-hentsch commented:
Based on the latest findings of https://github.com/SovereignCloudStack/issues/issues/229 I made the following changes/additions:

- only make the `dns` driver mandatory (not `dns_domain_ports`); this should be more then sufficient for internal DNS
- for infrastructures offering DNSaaS (Designate), instead make `dns_domain_ports` mandatory and `subnet_dns_published_fixed_ip` recommended

markus-hentsch commented:
I implemented a test script that verifies the existence of the API extensions as mandated by the current standard draft.
The script queries the Neutron Extensions API and relies on the information it provides.

However, due to https://bugs.launchpad.net/neutron/+bug/2063669 the test will succeed in *any* OVN-based setup since the DNS extensions are always reported as being available even if none of them actually are.

Upstream needs to fix this for the test script to actually report accurate results ...

anjastrunk commented on Standards/scs-01xx-v1-dns.md:
```diff
@@ -0,0 +1,167 @@
+
+However, this would also require solid DNS expertise at CSP-side to properly set up and integrate Designate and DNS zones as Designate does not act as a full DNS server on its own but instead relies on external DNS providers or self-hosted DNS infrastructures that the CSP needs to integrate into it.
+
+#### Mandating the use of DNSSEC
```
I doubt, that DNSSEC is common to every reader. Maybe it is worth to add at least a link to DNSSEC.

anjastrunk commented on Standards/scs-01xx-v1-dns.md:
```diff
@@ -0,0 +1,167 @@
+
+### Forwarded DNS
+
+- In OVS-based setups, the `dnsmasq_local_resolv` setting for Neutron DHCP agents MUST be disabled.
```
Please explain abbreviation "OVS".

anjastrunk reviewed (approved):
LGTM, beside missing two explanations for "OVM" and "OVS".

markus-hentsch commented:
> LGTM, beside missing two explanations for "OVM" and "OVS".

Thanks for pointing that out! This is quite important. I've added them to the glossary.

fkr commented on Standards/scs-01xx-v1-dns.md:
```diff
@@ -0,0 +1,122 @@
+
+The decisions of this standard will be categorized into three main sections: forwarded DNS, internal DNS and external DNS.
+
+Forwarded DNS refers to the DNS servers communicated to tenant VMs for DNS resolution as well as any local recursors involved.
```
@horazont "Announced DNS"?

horazont replied:
An entirely new term where I wouldn't know when reading it what it actually means.

Maybe "Default DNS servers" is what describes it best?

artificial-intelligence replied:
"forwarding" in DNS has a special defined meaning, please don't invent your own, to cite RFC 9499:

> Forwarding: The process of one server sending a DNS query with the RD bit set to 1 to another server to resolve that query. Forwarding is a function of a DNS resolver; it is different than simply blindly relaying queries.

> Forwarder: [Section 1](https://rfc-editor.org/rfc/rfc2308#section-1) of [[RFC2308](https://datatracker.ietf.org/doc/html/rfc2308)] describes a forwarder as "a nameserver used to resolve queries instead of directly using the authoritative nameserver chain". [[RFC2308](https://datatracker.ietf.org/doc/html/rfc2308)] further says "The forwarder typically either has better access to the internet, or maintains a bigger cache which may be shared amongst many resolvers." That definition appears to suggest that forwarders normally only query authoritative servers. In current use, however, forwarders often stand between stub resolvers and recursive servers. [[RFC2308](https://datatracker.ietf.org/doc/html/rfc2308)] is silent on whether a forwarder is iterative-only or can be a full-service resolver.

https://datatracker.ietf.org/doc/html/rfc9499#section-6-4.58

So please, do not invent your own terminology here, thanks.

horazont replied:
@artificial-intelligence Thank you for digging up the standard reference. This supports the claim that we need to use a term which is not related to "Forwarding".

(In OVN, there is no forwarder as er RFC9499 terminology.)

markus-hentsch replied:
> An entirely new term where I wouldn't know when reading it what it actually means.
> 
> Maybe "Default DNS servers" is what describes it best?

Since we seem to agree that "Forwarding" is a really misleading/incorrect choice here and no other suggestions have been made, I went with "Default DNS servers" for now. Feel free to comment other suggestions if any come to mind.

fkr commented on Standards/scs-01xx-v1-dns.md:
```diff
@@ -0,0 +1,170 @@
+
+#### Making Designate mandatory
+
+To offer a consistent feature set to customers, the SCS project could consider to make Designate mandatory in a sense that SCS clouds would need to integrate the service, make it available to customers and properly configure it for publishing DNS records.
```
```suggestion
To offer a consistent feature set to customers, the SCS project could consider to make Designate mandatory in a sense that SCS-compliant clouds would need to integrate the service, make it available to customers and properly configure it for publishing DNS records.
```

horazont commented on Standards/scs-01xx-v1-dns.md:
```diff
@@ -0,0 +1,122 @@
+
+- One or more local DNS recursors SHOULD be integrated into the infrastructure and the `dnsmasq_dns_server` setting SHOULD point to the local DNS recursor(s) only.
+    - Any local DNS recursor referenced by the `dnsmasq_dns_server` setting MUST implement DNSSEC validation.
+- If the cloud infrastructure has any provider networks connected to the internet, then the `dnsmasq_dns_server` entries MUST contain DNS servers (recursors or resolvers) that can resolve public DNS records.
```
This still needs some discussion, has that happened in some meeting I wasn't part of or should we schedule something for that? @markus-hentsch @fkr

artificial-intelligence commented on Standards/scs-01xx-v1-dns.md:
```diff
@@ -0,0 +1,122 @@
+
+A local DNS recursor can be used to cache and serve DNS responses locally. It servers as a proxy between the clients and external DNS servers.
+This improves performance and speed of DNS resolution in the infrastructure.
+Furthermore, it can be configured to use DNSSEC, DNS over HTTPS and/or DNS over TLS to increase security and privacy of DNS requests it handles for clients.
```
The security aspect has merit but DNS over HTTPS has significant different advantages over DNS over TLS, especially with regards to privacy and simplicity of configuration (e.g. port 443 is almost always no problem to connect to, good luck with DNSoTLS with some network middleboxes.)

horazont replied:
Which privacy advantages does DoH have which DoTLS does not have?

artificial-intelligence replied:
It's usage is harder to detect than DoT, this is why every browser implements DoH instead of DoT. This might or might not be a concern for our usecase though (I think it is the latter, but I didn't really analyze this in detail).

Specifically DoT uses Port 853 which makes it very easy to detect and block.

horazont replied:
Right, none of that matters for a recursor running locally in a datacenter. I'd argue even DoTLS is overkill for that, but that's not a hill I'm willing to die on.

artificial-intelligence reviewed (changes requested):
I need to review this in more detail later.
I'm only somewhat competent in the DNS RFCs as I happened to run, architect and maintain some dns infrastructure in the past.

DNS is a very complicated topic and I need some more time to read the standard very carefully, thanks.

markus-hentsch commented on Standards/scs-01xx-v1-dns.md:
```diff
@@ -0,0 +1,170 @@
+The DNSSEC extension to DNS ensures authenticity and integrity of the data provided to DNS resolvers.
+This protects against attacks like cache poisoning and increases the trustworthiness of the DNS infrastructure.
+
+This is a direct improvement for customers and yields low complexity for the CSP to implement since it is supported by most major DNS implementations natively.
```
Thanks for sharing your view on this. This is very valuable!

Do you consider the challenges with DNSSEC usage big enough to demote this to a recommendation?
Currently the standard 1)  makes DNSSEC mandatory for any local recursors and 2) mandates the use of public DNS servers that offer DNSSEC if no local recursor is used.

horazont replied:
@artificial-intelligence This text does not say you should run a DNSSEC auth.

This text only says that you should deploy a DNSSEC recursor, which AFAIK has none of the organisational complexities you mention.

artificial-intelligence replied:
It still has, of course, a lot of inert local complexity, e.g. the CSP will be the first point of contact for any customer who has dnssec related problems and must be able to deal with them.

That means you need trained support representatives which can debug and detect dnssec failures, e.g. when customers misconfigure their external domains because they don't know better.

Also there are still some subtle implementation bugs, e.g. some resolvers don't really properly support dnssec or only support it since $recent_version.

So I'm not really against mandating dnssec, you know, but I would like to see an honest analysis what the up- and downsides are and imho the current analysis is very superficially done on only a high level imho.

For example:
DNSSEC needs the availability of TCP Port 53 (strictly speaking you also need this for non-dnssec supporting resolvers, but it's way easier to get away with problems in your TCP based DNS setup if you don't do dnssec).

The default (forwarding) dns resolver in ml2/ovs is dnsmasq, which is not designed for large networks, it has no ability to configure it's available connection limit for TCP based dns replies. If you want to raise this, you have to recompile dnsmasq.

Further limitations apply to dnsmasq as well:
https://github.com/liquidm/dnsmasq/blob/master/FAQ#L31

artificial-intelligence replied:
I'll resolve this after todays discussion in the IaaS Breakout room, as the complexity of dnssec is only concentrated on a resolver provided by the CSP in this case, which enables customers of the CSP to work with dnssec.

artificial-intelligence commented on Standards/scs-01xx-v1-dns.md:
```diff
@@ -0,0 +1,122 @@
+| network | OpenStack Neutron network resource unless stated otherwise |
+| port | OpenStack Neutron port resource unless stated otherwise |
+| DNS | Domain Name System |
+| DNS recursor | Recursive DNS resolver |
```
you might want to refer to this section for this, where it is defined:

https://datatracker.ietf.org/doc/html/rfc9499#dns-servers-and-clients

HTH

markus-hentsch replied:
> you might want to refer to this section for this, where it is defined:
> 
> https://datatracker.ietf.org/doc/html/rfc9499#dns-servers-and-clients
> 
> HTH

Excellent suggestion and resource. Added. Thanks!

mbuechse commented on Standards/scs-01xx-v1-dns.md:
```diff
@@ -0,0 +1,170 @@
+---
+title: DNS Standard
+type: Standard
+status: Draft
```
The document should begin its life in `Proposal` state.

markus-hentsch replied:
Given the elaborate discussion that already happened in the corresponding issue, this PR as well as various community calls and breakout sessions, I think we are well beyond `Proposal` already.

Furthermore, I think the current process[^1] is not very clear when the transition between `Proposal` and `Draft` is supposed to happen. As I read it, the PR must be in `Draft` before the very first PR/merge anyway. But that's another topic.

[^1]: https://github.com/SovereignCloudStack/standards/blob/90f3244e4776a6d5a33b6402c5550826574a32bc/Standards/scs-0001-v1-sovereign-cloud-standards.md#process

artificial-intelligence reviewed (changes requested):
imho the standard should mandate RFC compliant DNS Servers that should be made available to the customer, given the fact there are a lot of non compliant dns servers out there.

horazont commented:
> imho the standard should mandate RFC compliant DNS Servers that should be made available to the customer, given the fact there are a lot of non compliant dns servers out there.

Good idea. Do you have a list of RFCs a good recursor should adhere to? I'm not sure if we need to list the RFCs of all RRs which we expect to be supported or whether that's something any base-RFC-compliant recursor _MUST_ handle correctly anyway (even though history has shown they don't always).

horazont commented on Standards/scs-01xx-v1-dns.md:
```diff
@@ -0,0 +1,170 @@
+To offer a consistent feature set to customers, the SCS project could consider to make Designate mandatory in a sense that SCS-compliant clouds would need to integrate the service, make it available to customers and properly configure it for publishing DNS records.
+This would offer easy DNS-as-a-Service functionality to customers.
+
+However, this would also require solid DNS expertise at CSP-side to properly set up and integrate Designate and DNS zones as Designate does not act as a full DNS server on its own but instead relies on external DNS providers or self-hosted DNS infrastructures that the CSP needs to integrate into it.
```
I think the levels of skills to run a proper auth vs. the set of skills required to run a recursor which is only reachable from the inside of your cloud are different by a couple orders of magnitude.

markus-hentsch commented:
On 2024-09-23 the current guidelines of the standard were discussed with CSP representatives (artcodix, plusserver, among others) in the Lean SCS Operator Coffee.
Below is a summary of the discussion points that I noted down:

- Neutron plugins for automated DNS record registration
    - there are problems with the internal DNS in conjunction with multi-tenant setups where the automatic naming is not powerful enough to prevent all collisions
    - even the external DNS focused plugins like "subnet-dns-publish-fixed-ip" and "dns-integration-domain-keywords" led to trouble with multi-region setups and collisions
    - in short, none of the Neutron integrations were without issues, especially in multi-tenant/multi-region setups; the CSPs eventually decided to disable all of them
- implementing local DNS recursors
    - it was argued that the vast majority of customers bring their own DNS configuration anyway, which leaves little incentive for a CSP to provide DNS servers themselves
    - in the eyes of the CSPs the standard would need to be extremely flexible and lenient when mandating any kind of local DNS implementation by the CSP
- DNSSEC
    - artcodix said that DNSSEC has not been of interest for any customers so far, so there was little motivation to implement it
    - plusserver reported that there were issues with DNSSEC in conjunction with their central DNS and anycast distribution to individual clouds in the past
    - although the challenge in actually implementing DNSSEC was rated to be rather small, the technical necessity of mandating DNSSEC via the standard was questioned by the CSPs

Unfortunately, this does not seem to leave a whole lot of options for the standard to include anything more than a few SHOULDs right now.

markus-hentsch commented:
I demoted some of the guidelines in this standard from MUST to SHOULD as a result of the recent CSP discussion (see [here](https://github.com/SovereignCloudStack/issues/issues/229#issuecomment-2379301726)).

As a result, the API tests I implemented for the conformance tests are not applicable anymore and I removed them. They [wouldn't have worked properly](https://bugs.launchpad.net/neutron/+bug/2063669) with Neutron's current implementation anyway.

artificial-intelligence reviewed (approved):
I like the addition of threats like zone squatting.

fkr commented:
@mbuechse I think, this should be merged now!
